Marlin/M99_Free_Mem_Chk.cpp

```diff
@@ -0,0 +1,277 @@
+		x = code_value();
+  		SERIAL_ECHOLNPGM("Corrupting free memory block.\n");
+  		ptr = (unsigned char *) __brkval;
+  		SERIAL_ECHOPAIR("\n__brkval : ",(long) ptr );
+  		ptr += 8;
+
+		sp = top_of_stack();
+  		SERIAL_ECHOPAIR("\nStack Pointer : ",(long) sp );
+  		SERIAL_ECHOLNPGM("\n");
+
+		n = sp - ptr - 64;  	// -64 just to keep us from finding interrupt activity that
+	       				// has altered the stack.
+		j = n / (x+1);
+		for(i=1; i<=x; i++) {
+			*(ptr+(i*j)) = i;
+  			SERIAL_ECHO("\nCorrupting address: 0x");
+		      	prt_hex_word( (unsigned int)  (ptr+(i*j)) );
+		}
+  		SERIAL_ECHOLNPGM("\n");
+		return;
+	}
+#endif
+
+//
+// M99 I    Initializes the free memory pool so it can be watched and prints vital
+// statistics that define the free memory pool.
+//
+	if (m99_not_initialized || code_seen('I') ) {				// If no sub-command is specified, the first time
+  		SERIAL_ECHOLNPGM("Initializing free memory block.\n");   	// this happens, it will Initialize.
+  		ptr = (unsigned char *) __brkval;				// Repeated M99 with no sub-command will not destroy the
+  		SERIAL_ECHOPAIR("\n__brkval : ",(long) ptr );			// state of the initialized free memory pool.
+  		ptr += 8;
+
+		sp = top_of_stack();
+  		SERIAL_ECHOPAIR("\nStack Pointer : ",(long) sp );
+  		SERIAL_ECHOLNPGM("\n");
+
+		n = sp - ptr - 64;  	// -64 just to keep us from finding interrupt activity that
+	       				// has altered the stack.
+
+  		SERIAL_ECHO( n );
+  		SERIAL_ECHOLNPGM(" bytes of memory initialized.\n");
+
+		for(i=0; i<n; i++)
+			*(ptr+i) = (unsigned char) 0xe5;
+
+		for(i=0; i<n; i++) {
+			if ( *(ptr+i) != (unsigned char) 0xe5 ) {
+  				SERIAL_ECHOPAIR("? address : ", (unsigned long) ptr+i );
+  				SERIAL_ECHOPAIR("=", *(ptr+i) );
+  				SERIAL_ECHOLNPGM("\n");
+			}
+		}
+		m99_not_initialized = 0;
+  		SERIAL_ECHOLNPGM("Done.\n");
+		return;
+	}
+	return;
+}
+
+// top_of_stack() returns the location of a variable on its stack frame.  The value returned is above
+// the stack once the function returns to the caller.
+
+unsigned char *top_of_stack() {
+unsigned char x;
+	return &x;
+}
+
+//
+// 3 support routines to print hex numbers.  We can print a nibble, byte and word
+//
+
+void prt_hex_nibble( unsigned int n )
+{
+	if ( n <= 9 )
+		SERIAL_ECHO(n);
+	else
+		SERIAL_ECHO( (char) ('A'+n-10) );
+	delay(2);
+}
+
+void prt_hex_byte(unsigned int b)
+{
+	prt_hex_nibble( ( b & 0xf0 ) >> 4 );
+	prt_hex_nibble(  b & 0x0f );
+}
+
+void prt_hex_word(unsigned int w)
+{
+	prt_hex_byte( ( w & 0xff00 ) >> 8 );
+	prt_hex_byte(  w & 0x0ff );
+}
+
+// how_many_E5s_are_here() is a utility function to easily find out how many 0xE5's are
+// at the specified location.  Having this logic as a function simplifies the search code.
+//
+int how_many_E5s_are_here( unsigned char *p)
+{
+int n;
+
+	for(n=0; n<32000; n++) {
+		if ( *(p+n) != (unsigned char) 0xe5)
+			return n-1;
+	}
+	return -1;
+}
+
+#endif
+
```

Comment:
Instead of tabs use 2 spaces. http://www.marlinfirmware.org/index.php/Coding_Standards